mne/preprocessing/flat.py

```diff
@@ -0,0 +1,116 @@
+    %(picks_good_data)s
+    %(verbose)s
+
+    Returns
+    -------
+    raw : instance of Raw
+        The modified raw instance. Operates in place.
+
+    Notes
+    -----
+    For a given channel :math:`i` in `picks` that is flat (considering only
+    flat segments of at least `min_duration`)
+    for :math:`p_{i,t}` proportion of time points, given the spatial fraction
+    of picks that the channel represents (constant
+    :math:`p_{i,s} = 1 / N_\mathrm{picks}`), the `ratio`
+    parameter sets the ratio :math:`\frac{p_{i,t}}{p_{i,s}}` below which
+    temporal marking (:class:`~mne.Annotations`) will be used instead of
+    spatial marking (:clasS:`~mne.Info`). Or more simply:
+
+    - ``ratio < 1. / len(picks)`` always marks temporally.
+        - This risks the entire recording being marked bad if there is
+          (at least) one channel that is flat at all times.
+    - ``1. / len(picks) <= ratio < len(picks)`` marks spatially or temporally.
+        - Values in the range ``0 < ratio < 1`` prefer temporal marking to
+          spatial marking.
+        - The default value ``ratio = 1.`` uses the temporal method
+          if and only if the proportion of time a given channel is flat is
+          less than the reciprocal of the number of channels in ``picks``.
+        - Values in the range ``1 < ratio < len(picks)`` prefer spatial
+          marking to temporal marking.
+    - ``ratio >= len(picks)`` always marks spatially.
+        - This risks a channel being marked as bad for the entire recording if
+          it is flat for a contiguous `min_duration` chunk of time.
+
+    .. note:: This function may perform much faster if data are loaded
+              in memory, as it loads data one channel at a time (across all
+              time points), which is typically not an efficient way to read
+              raw data from disk.
+
+    .. versionadded:: 0.18
+    """
+    _validate_type(raw, BaseRaw, 'raw')
+    ratio = float(ratio)
+    min_duration = float(min_duration)
+    picks = _picks_to_idx(raw.info, picks, 'data_or_ica', exclude='bads')
+    # This will not be so efficient for most readers, but we can optimize
+    # it later
+    any_flat = np.zeros(len(raw.times), bool)
+    bads = list()
+    time_thresh = int(np.round(min_duration * raw.info['sfreq']))
+    with ProgressBar(picks, mesg='Finding flat segments', spinner=True,
+                     verbose_bool='auto') as pb:
+        for pick in pb:
+            data = raw[pick][0][0]
+            flat = np.diff(data) == 0
+            flat = np.concatenate(
+                [flat[[0]], flat[1:] | flat[:-1], flat[[-1]]])
+            starts, stops = _mask_to_onsets_offsets(flat)
+            for start, stop in zip(starts, stops):
+                if stop - start < time_thresh:
+                    flat[start:stop] = False
+            flat_mean = flat.mean()
+            with np.errstate(divide='ignore'):
+                bad_ratio = np.divide(1., flat_mean * len(picks))
```

Comment:
this parametrization is weird to me. Why not just comparing flat_mean to ratio. If flat_mean > 0.3 then you have more than 30% of time samples for which it's flat.

Reply:
The idea is that `ratio` should have to do with the loss of spatial information vs the loss of temporal information

Reply:
But I guess you could easily manually convert to this trade-off format knowing the length of `picks`?

Reply:
It made the API quite a bit clearer I think. And now the default actually works pretty well for my dataset (before I had to tweak it).